# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [3]:
import pandas as pd
df=pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')
df

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [6]:
#1.1
df.shape

(4008, 11)

In [10]:
#1.2
df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

#1.2 - Comments on the right side
Customer                      object - Correct
ST                            obje  - Correctt
GENDER                        obje - Correctct
Education                     obj - Correctect
Customer Lifetime Value       ob - Better as a floatject
Income                       fl - Correctoat64
Monthly Premium Auto         f - Correctloat64
Number of Open Complaints     - Better as an integer object
Policy Type                  - Correct  object
Vehicle Class               - Correct   object
Total Claim Amount         - Correctype: object


In [11]:
#1.3
df.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [15]:
#print(f"CLV max {df['Customer Lifetime Value'].max()} and min. {df['Customer Lifetime Value'].min()}")
print(f"Income max {df['Income'].max()} and min. {df['Income'].min()}")
print(f"Monthly Premium Auto max {df['Monthly Premium Auto'].max()} and min. {df['Monthly Premium Auto'].min()}")
print(f"Total Claim Amount max {df['Total Claim Amount'].max()} and min. {df['Total Claim Amount'].min()}")

Income max 99960.0 and min. 0.0
Monthly Premium Auto max 35354.0 and min. 61.0
Total Claim Amount max 2893.239678 and min. 0.382107


#1.3
Customer                     1071
ST                               - Categorical, nothing to say until a more detailed observation is done, like typing errors control duplicating States.8
GENDER                          - Categorical, the number should be 2, which means that some data cleaning should be done. 5
Education                      - Categorical, nothing to say until a more detailed observation is done  6
Customer Lifetime Value       - Should be Numerical. There are more CLV than customer IDs, which probably mean that a more careful study and correction should be executed.1027
Income                       - Numerical. range: Income max 99960.0 and min. 0.0  774
Monthly Premium Auto        - Numerical range: Monthly Premium Auto max 35354.0 and min. 61.0. A value of 35354 it's very different from the minimum, should be checked if it's correct and compared with others.   132
Number of Open Complaints  - Should be numerical, since it represents a quantity ("number of..").      6
Policy Type               - Categorical       3
Vehicle Class            - Categorical        6
Total Claim Amount      - Numerical       761

In [36]:
#4 mean, median, mode, standard deviation, and quartiles
df.describe()

Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

In [34]:
#4 finding mode()
print(df["Income"].mode())
print(df["Monthly Premium Auto"].mode())
print(df["Total Claim Amount"].mode())
#0    0.0
#Name: Income, dtype: float64
#0    65.0
#Name: Monthly Premium Auto, dtype: float64
#0    321.6
#Name: Total Claim Amount, dtype: float64

0    0.0
Name: Income, dtype: float64
0    65.0
Name: Monthly Premium Auto, dtype: float64
0    321.6
Name: Total Claim Amount, dtype: float64


#4 Conclusion for numericals
Income: the median if quite near to the mean, but the STD if quite high, meaning the data is probably disperse specially in the last quartile (75% up).
Monthly Premium Auto: the median is highly discrepant form the mean, which indicates that some clients wouyld be paying significantly more - it was adressed previously that it could be an error (data to be checked). The most is nearer to the median, which supports what was previously mentioned. Very assimetrical distribution.
Total Claim Amount: A few claims in the last quartile increasing the mean in comparison with the median. STD shows big dispersion. There is not any client without a claim value.

In [79]:
print(df["ST"].mode())
print(df["GENDER"].mode())
print(df["Education"].mode())
print(df["Policy Type"].mode())
print(df["Vehicle Class"].mode())

0    California
Name: ST, dtype: object
0    F
Name: GENDER, dtype: object
0    Bachelor
Name: Education, dtype: object
0    Personal Auto
Name: Policy Type, dtype: object
0    Four-Door Car
Name: Vehicle Class, dtype: object


array(['Washington', 'Arizona', 'Nevada', 'California', 'Oregon', nan],
      dtype=object)

In [61]:
print(df['ST'].value_counts(), df['ST'].value_counts(normalize=True)*100, "\n")
print(df['GENDER'].value_counts(), df['GENDER'].value_counts(normalize=True)*100, "\n")
print(df['Education'].value_counts(), df['Education'].value_counts(normalize=True)*100, "\n")
print(df["Policy Type"].value_counts(), df['Policy Type'].value_counts(normalize=True)*100, "\n")
print(df["Vehicle Class"].value_counts(), df['Vehicle Class'].value_counts(normalize=True)*100, "\n")

ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64 ST
Oregon        29.878618
California    19.701214
Arizona       17.366947
Cali          11.204482
Nevada         9.150327
Washington     7.563025
WA             2.801120
AZ             2.334267
Name: proportion, dtype: float64 

GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64 GENDER
F         47.903564
M         43.291405
Male       4.088050
female     2.935010
Femal      1.781971
Name: proportion, dtype: float64 

Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64 Education
Bachelor                30.252101
College                 29.225023
High School or Below    27.637722
Master                   8.776844
Do

#5 Conclusion for numericals
ST: California (~31%), Oregon (~29%), Arizona (~19%) the 3 main states that the customers are from. Data must be cleaned (names like "Cali" or abrevitions like "AZ" for arizona must be changed, as an example)
GENDER: very proportional.
Education: Only 12% with a high educational level, maybe relatable with the income discrepancies among customers.
Policy type: 70% of customers have private and 20% corporate car policies
Vehicle class: Four-Door Car (54%), Two-Door Car (19%) and SUV (18%) are the most insured car types.

ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64 ST
Oregon        29.878618
California    19.701214
Arizona       17.366947
Cali          11.204482
Nevada         9.150327
Washington     7.563025
WA             2.801120
AZ             2.334267
Name: proportion, dtype: float64 

GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64 GENDER
F         47.903564
M         43.291405
Male       4.088050
female     2.935010
Femal      1.781971
Name: proportion, dtype: float64 

Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64 Education
Bachelor                30.252101
College                 29.225023
High School or Below    27.637722
Master                   8.776844
Doctor                   3.454715
Bachelors                0.653595
Name: proportion, dtype: float64 

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64 Policy Type
Personal Auto     72.829132
Corporate Auto    21.848739
Special Auto       5.322129
Name: proportion, dtype: float64 

Vehicle Class
Four-Door Car    576
Two-Door Car     205
SUV              199
Sports Car        57
Luxury SUV        20
Luxury Car        14
Name: count, dtype: int64 Vehicle Class
Four-Door Car    53.781513
Two-Door Car     19.140990
SUV              18.580766
Sports Car        5.322129
Luxury SUV       1.867414
Luxury Car        1.307190
Name: proportion, dtype: float64 


## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [77]:
#df['ST']=df['ST'].replace("Cali", "California")
#df["ST"]=df["ST"].replace("AZ", "Arizona", and "WA", "Washington")
STreplacements= {
    "AZ": "Arizona",
    "WA": "Washington",
    'Cali': 'California'
}
df["ST"]=df["ST"].replace(STreplacements)
STpercentage= df['ST'].value_counts(normalize=True)*100
STSort=STpercentage.sort_values(ascending=True)
STSort.head(5)

ST
Nevada         9.150327
Washington    10.364146
Arizona       19.701214
Oregon        29.878618
California    30.905696
Name: proportion, dtype: float64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [94]:
print(f'The total number of policies sold were {df["Policy Type"].count().sum()}, corresponding to each one: {df["Policy Type"].value_counts()}')

The total number of policies sold were 1071, corresponding to each one: Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [124]:
PAdf= df.loc[df['Policy Type'] == "Personal Auto"]
PCdf= df.loc[df['Policy Type'] == "Corporate Auto"]

print(f"Corporate Customer: {PCdf['Income'].mean()} \nPersonal CustomerCorporate Customer: {PAdf['Income'].mean()}\n")

Corporate Customer: 41390.31196581197 
Personal CustomerCorporate Customer: 38180.69871794872



### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [130]:
HighCAdf=df[df['Total Claim Amount']>532.800000]
HighCAdf.describe()

Income  Monthly Premium Auto  Total Claim Amount
count    264.000000            264.000000          264.000000
mean   23677.344697            165.193182          782.228263
std    27013.483721            623.930992          292.751640
min        0.000000             63.000000          537.600000
25%        0.000000             99.000000          606.521741
50%    18807.000000            114.000000          679.597985
75%    42423.750000            133.250000          851.400000
max    99316.000000          10202.000000         2893.239678

In [132]:
print(HighCAdf['ST'].value_counts(), HighCAdf['ST'].value_counts(normalize=True)*100, "\n")
print(HighCAdf['GENDER'].value_counts(), HighCAdf['GENDER'].value_counts(normalize=True)*100, "\n")
print(HighCAdf['Education'].value_counts(), HighCAdf['Education'].value_counts(normalize=True)*100, "\n")
print(HighCAdf["Policy Type"].value_counts(), HighCAdf['Policy Type'].value_counts(normalize=True)*100, "\n")
print(HighCAdf["Vehicle Class"].value_counts(), HighCAdf['Vehicle Class'].value_counts(normalize=True)*100, "\n")

ST
Oregon        90
California    86
Arizona       49
Nevada        29
Washington    10
Name: count, dtype: int64 ST
Oregon        34.090909
California    32.575758
Arizona       18.560606
Nevada        10.984848
Washington     3.787879
Name: proportion, dtype: float64 

GENDER
F         115
M         100
Male       16
female      5
Femal       2
Name: count, dtype: int64 GENDER
F         48.319328
M         42.016807
Male       6.722689
female     2.100840
Femal      0.840336
Name: proportion, dtype: float64 

Education
Bachelor                85
High School or Below    85
College                 82
Master                   8
Doctor                   4
Name: count, dtype: int64 Education
Bachelor                32.196970
High School or Below    32.196970
College                 31.060606
Master                   3.030303
Doctor                   1.515152
Name: proportion, dtype: float64 

Policy Type
Personal Auto     191
Corporate Auto     60
Special Auto       13
Name: count, dtype: